manifests/jail.pp

```diff
@@ -1,64 +1,147 @@
+#   Useful for shared jails. Defaults to the value of group.
+#
+# @param password_authentication
+#   Can the user login with a password? Public key authentication is generally
+#   recommended and has to be configured outside of the scope of this module.
+#
 define sftp_jail::jail (
-  $jail_name               = $name,
-  $user                    = undef,
-  $group                   = undef,
-  $match_group             = undef,
-  $password_authentication = 'no',
+  Enum['present', 'absent'] $ensure                  = 'present',
+  Sftp_jail::File_name      $jail_name               = $name,
+  Stdlib::Absolutepath      $jail_base               = "${sftp_jail::chroot_base}/${jail_name}",
+  Accounts::User::Name      $user                    = $name,
+  Accounts::User::Name      $group                   = $user,
+  Sftp_jail::Groups         $groups                  = [],
+  Optional[String[100]]     $password_hash           = undef,
+  Stdlib::Absolutepath      $real_home               = "${jail_base}/home/${user}",
+  Stdlib::Absolutepath      $home                    = "/home/${user}",
+  Optional[String[2]]       $comment                 = undef,
+  Boolean                   $manage_user             = false,
+  Boolean                   $sftp_admin              = false,
+  Sftp_jail::Sub_dirs       $sub_dirs                = $sftp_jail::sub_dirs,
+  Boolean                   $merge_subdirs           = false,
+  Boolean                   $purge                   = false,
+  Accounts::User::Name      $match_group             = $group,
+  Boolean                   $password_authentication = false,
 ) {
   include sftp_jail
-  $jail_base = "${sftp_jail::chroot_base}/${jail_name}"
 
-  if ($match_group) {
-    $ssh_match_group = $match_group
-  }
-  else {
-    $ssh_match_group = $group
-  }
-
-  if !($group) {
-    $group = $user
-  }
-
-  file { $jail_base:
-    ensure  => 'directory',
-    owner   => 'root',
-    group   => 'root',
-    mode    => '0755',
-    require => File[$sftp_jail::chroot_base],
-  }
-
-  file { "${jail_base}/incoming":
-    ensure  => 'directory',
-    owner   => $user,
-    group   => $group,
-    mode    => '0775',
-    require => File[$sftp_jail::chroot_base],
-  }
-
-  file { "${jail_base}/home":
-    ensure  => 'directory',
-    owner   => 'root',
-    group   => 'root',
-    mode    => '0755',
-    require => File[$sftp_jail::chroot_base],
+  unless $sftp_admin or (($ensure == 'absent') and !$purge) {
+    if $ensure == 'absent' {
+      $_ensure = 'absent'
+      $force = true
+    } else {
+      $_ensure = 'directory'
+      $force = undef
+    }
+    file { default:
+        ensure => $_ensure,
+        owner  => 'root',
+        group  => 'root',
+        mode   => '0755',
+        force  => $force,
+      ;
+      $jail_base: ;
+      "${jail_base}/home":;
+      "${jail_base}/incoming":
+        owner => $user,
+        group => $group,
+        mode  => '0775',
+      ;
+    }
```

Comment:
I personally prefer the old style and dislike the defaults. If you want, you can combine the root owned directories
```puppet
file { [$jail_base, "${jail_base}/home"]:
  ...
}
```

Changing the style and functionality in the same PR also make it harder for a reviewer to review.